pulsar-client-reactive-api/src/test/java/org/apache/pulsar/reactive/client/internal/api/InflightLimiterTest.java

```diff
@@ -16,50 +16,174 @@
+					});
+				}).doOnRequest((requested) -> {
+					worker.schedule(() -> {
+						currentRequests.addAndGet(requested);
+						long current = totalRequests.addAndGet(requested);
+						requestsMax.accumulateAndGet(current, Math::max);
+					});
+				}).doOnNext((__) -> {
+					worker.schedule(() -> {
+						currentRequests.decrementAndGet();
+						totalRequests.decrementAndGet();
+						subscriptionsActiveBeforeCompletingFirstElement.decrementAndGet();
+						// Mono will complete after this element, so reduce all remaining
+						// requests;
+						totalRequests.addAndGet(-currentRequests.getAndSet(0));
+					});
+				}).transform(inflightLimiter::transform).subscribeOn(Schedulers.single());
+			}, maxElements);
+			List<Integer> values = flux.collectList().block();
+			SoftAssertions.assertSoftly((softly) -> {
+				softly.assertThat(values).containsExactlyInAnyOrderElementsOf(inputValues);
+				softly.assertThat(requestsMax.get()).as("requestsMax").isEqualTo(maxInflight);
+				softly.assertThat(subscriptionsMax.get()).as("subscriptionsMax").isEqualTo(maxInflight);
+			});
+		}
+		finally {
+			worker.dispose();
+		}
+	}
+
+	@ParameterizedTest
+	@CsvSource({ "7,100", "13,100", "37,500", "51,1000" })
+	void shouldNotSubscribeMoreThanMaxInflightForMonos(int maxInflight, int maxElements) {
+		InflightLimiter inflightLimiter = new InflightLimiter(maxInflight, maxInflight, Schedulers.single(),
+				InflightLimiter.DEFAULT_MAX_PENDING_SUBSCRIPTIONS);
+
+		AtomicInteger completableFuturesInflight = new AtomicInteger();
+		AtomicInteger completableFuturesInflightMax = new AtomicInteger();
+
+		List<Integer> inputValues = IntStream.rangeClosed(1, maxElements).boxed().collect(Collectors.toList());
+
+		Flux<Integer> flux = Flux.fromIterable(inputValues).flatMap((i) -> Mono.fromCompletionStage(() -> {
+			completableFuturesInflightMax.accumulateAndGet(completableFuturesInflight.incrementAndGet(), Math::max);
+			CompletableFuture<Integer> future = new CompletableFuture<>();
+			Schedulers.boundedElastic().schedule(() -> {
+				completableFuturesInflight.decrementAndGet();
+				future.complete(i);
+			}, 5, TimeUnit.MILLISECONDS);
+			return future;
+		}).transform(inflightLimiter::transform).subscribeOn(Schedulers.parallel()), maxElements);
+
+		List<Integer> values = flux.collectList().block();
+		assertThat(values).containsExactlyInAnyOrderElementsOf(inputValues);
+		assertThat(completableFuturesInflightMax.get()).isEqualTo(maxInflight);
 	}
 
-	private void verifyFairness(List<Integer> values) {
-		int previousValue = 1;
-		for (int i = 0; i < values.size(); i++) {
-			int value = values.get(i) % 100;
-			if (value == 0) {
-				value = 100;
-			}
-			assertThat(Math.abs(previousValue - value)).as("value %d at index %d", values.get(i), i)
-					.isLessThanOrEqualTo(48);
-			previousValue = value;
+	@Disabled("This test was used to debug the flakiness issue")
+	@RepeatedTest(100)
+	void repeatShouldNotSubscribeMoreThanMaxInflightForMonos() {
```

Comment:
Maybe we could remove it now ?

Reply:
It is useful for running a stress test manually. I don't see much harm of keeping this around. It might become useful when making changes to InflightLimiter.